{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment in Bioinformatics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Needleman-Wunsch Algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Global sequence alignment** assumes two sequences are homologous (or share a common evolutionary ancestor) across their whole length. It tries to line up the whole thing, from tail to snout. This might be a reasonable assumption for two genes in the same gene family, for instance.\n",
    "\n",
    "First, we import some modules which needed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we create class `NeedlemanWunsch` to implement the Needleman-Wunsch algorithm in python:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NeedlemanWunsch:\n",
    "    \"\"\" This is a dynamic programming algorithm for finding the optimal\n",
    "    global alignment of two sequences.\n",
    "    \"\"\"\n",
    "    def __init__(self, seq1, seq2, gap_penalty=-1, match_award=1, mismatch_penalty=-1):\n",
    "        self.seq1, self.seq2 = seq1, seq2\n",
    "        self.M, self.N = len(self.seq1)+1, len(self.seq2)+1\n",
    "        self.gap_penalty = gap_penalty\n",
    "        self.match_award = match_award\n",
    "        self.mismatch_penalty = mismatch_penalty\n",
    "\n",
    "    def pretty_table_from_array(self, data_array, row_labels, col_labels):\n",
    "        \"\"\" Show an HTML table from a 2d numpy array \"\"\"\n",
    "        df = pd.DataFrame(data_array, index=row_labels, columns=col_labels)\n",
    "        table_html = df.to_html()\n",
    "        return HTML(table_html)\n",
    "\n",
    "    def match_score(self, alpha, beta):\n",
    "        \"\"\" Get the match score between two sequences \"\"\"\n",
    "        if alpha == beta:\n",
    "            return self.match_award\n",
    "        else:\n",
    "            return self.mismatch_penalty\n",
    "\n",
    "    def score_matrix(self):\n",
    "        \"\"\" Get the score matrix and traceback matrix between two sequences \"\"\"\n",
    "        # Initiation\n",
    "        score = np.zeros((self.M, self.N))\n",
    "        trace = np.full([self.M, self.N], '-')\n",
    "        # Define the arrows from different directions\n",
    "        up_arrow = \"\\u2191\"\n",
    "        left_arrow = \"\\u2190\"\n",
    "        up_left_arrow = \"\\u2196\"\n",
    "        # Calculate\n",
    "        # Fill out cell(0,0)\n",
    "        score[0, 0] = 0\n",
    "        trace[0, 0] = '-'\n",
    "        # Fill out first column\n",
    "        for i in range(1, self.M):\n",
    "            score[i, 0] = self.gap_penalty * i\n",
    "            trace[i, 0] = up_arrow\n",
    "        # Fill out first row\n",
    "        for j in range(1, self.N):\n",
    "            score[0, j] = self.gap_penalty * j\n",
    "            trace[0, j] = left_arrow\n",
    "        # Fill out all other cell\n",
    "        for i in range(1, self.M):\n",
    "            for j in range(1, self.N):\n",
    "                match = score[i-1, j-1] + self.match_score(self.seq1[i-1], self.seq2[j-1])\n",
    "                delete = score[i-1, j] + self.gap_penalty\n",
    "                insert = score[i, j-1] + self.gap_penalty\n",
    "                score[i, j] = max(match, delete, insert)\n",
    "                trace[i, j] = [up_left_arrow, left_arrow, up_arrow][np.argmax([match, delete, insert])]\n",
    "        return score, trace\n",
    "\n",
    "    def traceback(self):\n",
    "        \"\"\" Traceback and calculate the alignment \"\"\"\n",
    "        # Get the score array from def(score_matrix)\n",
    "        score, _ = self.score_matrix()\n",
    "        # Create variables to store alignment\n",
    "        align1 = \"\"\n",
    "        align2 = \"\"\n",
    "        # start from the bottom right cell\n",
    "        i, j = self.M-1, self.N-1\n",
    "        # Traceback\n",
    "        while i > 0 and j > 0:\n",
    "            # match or mismatch\n",
    "            if score[i, j] == score[i-1, j-1] + self.match_score(self.seq1[i-1], self.seq2[j-1]):\n",
    "                align1 += self.seq1[i-1]\n",
    "                align2 += self.seq2[j-1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            # delete\n",
    "            elif score[i, j] == score[i-1, j] + self.gap_penalty:\n",
    "                align1 += self.seq1[i-1]\n",
    "                align2 += '-'\n",
    "                i -= 1\n",
    "            # insert\n",
    "            elif score[i, j] == score[i, j-1] + self.gap_penalty:\n",
    "                align1 += '-'\n",
    "                align2 += self.seq2[j-1]\n",
    "                j -= 1\n",
    "        # Finish tracing\n",
    "        while i > 0:\n",
    "            align1 += self.seq1[i-1]\n",
    "            align2 += '-'\n",
    "            i -= 1\n",
    "        while j > 0:\n",
    "            align1 += '-'\n",
    "            align2 += self.seq2[j-1]\n",
    "            j -= 1\n",
    "        # Reverse the alignment\n",
    "        align1 = align1[::-1]\n",
    "        align2 = align2[::-1]\n",
    "        return align1, align2\n",
    "\n",
    "    def display_alignment(self):\n",
    "        \"\"\" Display the global alignment between two sequences \"\"\"\n",
    "        # Get the alignment result\n",
    "        align1, align2 = self.traceback()\n",
    "        # Get the map label index\n",
    "        map_label = [\" \"] * len(align1)\n",
    "        for i in range(len(align1)):\n",
    "            if align1[i] == align2[i]:\n",
    "                map_label[i] = \"|\"\n",
    "            elif align1[i] == '-' or align2[i] == '-':\n",
    "                continue\n",
    "            else:\n",
    "                map_label[i] = '.'\n",
    "\n",
    "        map_label = ''.join(map_label)\n",
    "        # Display the alignment\n",
    "        print(align1 + \"\\n\" + map_label + \"\\n\" + align2)\n",
    "\n",
    "    def display_array(self):\n",
    "        \"\"\" Display the score array and trace array \"\"\"\n",
    "        # Get the score array and trace array\n",
    "        score_array, traceback_array = self.score_matrix()\n",
    "        # Display\n",
    "        column_labels = [label for label in \"-\" + \"\".join(self.seq1)]\n",
    "        row_labels = [label for label in \"-\" + \"\".join(self.seq2)]\n",
    "        print('Score Matrix:')\n",
    "        display(self.pretty_table_from_array(score_array, row_labels, column_labels))\n",
    "        print('Traceback Matrix:')\n",
    "        display(self.pretty_table_from_array(traceback_array, row_labels, column_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's an example. Let's start with two random unaligned DNA sequences created by `numpy.random`:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)\n",
    "x = np.random.choice(['A', 'T', 'G', 'C'], 8)\n",
    "y = np.random.choice(['A', 'T', 'G', 'C'], 8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We set different gap penalty [0, -1, -2] to test the Needleman-Wunsch algorithm:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "nw0 = NeedlemanWunsch(x, y, gap_penalty=0)\n",
    "nw1 = NeedlemanWunsch(x, y, gap_penalty=-1)\n",
    "nw2 = NeedlemanWunsch(x, y, gap_penalty=-2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we check out the score matries and traceback matries, firstly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score Matrix:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>-6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-4.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-5.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-6.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-7.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-8.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback Matrix:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>-</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nw1.display_array()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The score matrix shown above can now be read as a map that show better or worse paths through the alignment of seq1 and seq2, with higher numebrs representing better paths. \n",
    "\n",
    "There are a couple of rules for interpreting the matrix. They may or may not be intuitive at first, but if we follow them we can reconstruct all the optimal alignments from the matrix. \n",
    "\n",
    "- We start at the lower right. Needleman-Wunsch assumes a **global alignment**, meaning the whole sequence is aligned. Therefore we start at the end and work backward.\n",
    "- Each horizontal step represents insertion of an indel('-') into the sequence of the rows.\n",
    "- Each verical step represents insertion of an indel('-') into the top sequence.\n",
    "- The optimal alignments are represented by the cells with the highest scores.\n",
    "\n",
    "Then, we check out the alignment between seq1 and seq2 under different gap penalty:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gap penalty = 0:\n",
      "---GCAGGCAA\n",
      "   |  ||| |\n",
      "GTGG--GGC-A\n",
      "\n",
      "gap penalty = -1:\n",
      "G-CAGGCAA\n",
      "| ..||| |\n",
      "GTGGGGC-A\n",
      "\n",
      "gap penalty = -2:\n",
      "GCAGGCAA\n",
      "|..||..|\n",
      "GTGGGGCA\n"
     ]
    }
   ],
   "source": [
    "print('gap penalty = 0:')\n",
    "nw0.display_alignment() # gap penalty = 0\n",
    "print(\"\\n\" + 'gap penalty = -1:')\n",
    "nw1.display_alignment() # gap penalty = -1\n",
    "print(\"\\n\" + 'gap penalty = -2:')\n",
    "nw2.display_alignment() # gap penalty = -2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "where, '|' represents match perfectly, '.' represents mismatch, and space represents indel between two sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Smith-Waterman Algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Local sequence alignment** assumes that there are regions of homology within a bigger sequence, but also regions that do not share common ancestry. This might be a reasonable assumption if checking how many genes two species share. If the species aren't extremely closely related, we would expect that they would share certain genes, but also that there should be large portions of the genome that do not share detectable evolutionary ancestry between the two species (due to large insertions, deletions, inversions or rearrangements).\n",
    "\n",
    "First, we import some modules needed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we create class `SmithWaterman` to implement the Smith-Waterman algorithm in python:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SmithWaterman:\n",
    "    \"\"\" This is a dynamic programming algorithm for finding the optimal\n",
    "    local alignment of two sequences.\n",
    "    \"\"\"\n",
    "    def __init__(self, seq1, seq2, gap_penalty=-1, match_award=1, mismatch_penalty=-1):\n",
    "        self.seq1, self.seq2 = seq1, seq2\n",
    "        self.M, self.N = len(self.seq1)+1, len(self.seq2)+1\n",
    "        self.gap_penalty = gap_penalty\n",
    "        self.match_award = match_award\n",
    "        self.mismatch_penalty = mismatch_penalty\n",
    "        self.max_score = -1\n",
    "        self.max_index = (-1, -1)\n",
    "\n",
    "    def pretty_table_from_array(self, data_array, row_labels, col_labels):\n",
    "        \"\"\" Show an HTML table from a 2d numpy array \"\"\"\n",
    "        df = pd.DataFrame(data_array, index=row_labels, columns=col_labels)\n",
    "        table_html = df.to_html()\n",
    "        return HTML(table_html)\n",
    "\n",
    "    def match_score(self, alpha, beta):\n",
    "        \"\"\" Get the match score between two sequences \"\"\"\n",
    "        if alpha == beta:\n",
    "            return self.match_award\n",
    "        else:\n",
    "            return self.mismatch_penalty\n",
    "\n",
    "    def score_matrix(self):\n",
    "        \"\"\" Get the score matrix between two sequences \"\"\"\n",
    "        # Initiation\n",
    "        score = np.zeros((self.M, self.N))\n",
    "        trace = np.full([self.M, self.N], '-')\n",
    "        # Define the arrows from different directions\n",
    "        up_arrow = \"\\u2191\"\n",
    "        left_arrow = \"\\u2190\"\n",
    "        up_left_arrow = \"\\u2196\"\n",
    "        # Calculate\n",
    "        # Fill out cell(0,0)\n",
    "        score[0, 0] = 0\n",
    "        trace[0, 0] = '-'\n",
    "        # Fill out first column\n",
    "        for i in range(1, self.M):\n",
    "            score[i, 0] = self.gap_penalty * i\n",
    "            trace[i, 0] = up_arrow\n",
    "        # Fill out first row\n",
    "        for j in range(1, self.N):\n",
    "            score[0, j] = self.gap_penalty * j\n",
    "            trace[0, j] = left_arrow\n",
    "        # Fill out all other cell\n",
    "        for i in range(1, self.M):\n",
    "            for j in range(1, self.N):\n",
    "                match = score[i-1, j-1] + self.match_score(self.seq1[i-1], self.seq2[j-1])\n",
    "                delete = score[i-1, j] + self.gap_penalty\n",
    "                insert = score[i, j-1] + self.gap_penalty\n",
    "                score[i, j] = max(0, match, delete, insert)\n",
    "                trace[i, j] = [up_left_arrow, left_arrow, up_arrow][np.argmax([match, delete, insert])]\n",
    "                # Get the best locally alignment\n",
    "                if score[i, j] >= self.max_score:\n",
    "                    self.max_index = (i, j)\n",
    "                    self.max_score = score[i, j]\n",
    "        return score, trace\n",
    "\n",
    "    def traceback(self):\n",
    "        \"\"\" Traceback and calculate the alignment \"\"\"\n",
    "        # Get the score matrix from def(score_matrix)\n",
    "        score, _ = self.score_matrix()\n",
    "        # Create variables to store alignment\n",
    "        align1 = \"\"\n",
    "        align2 = \"\"\n",
    "        # start from the best score cell\n",
    "        (i, j) = self.max_index\n",
    "        # Traceback\n",
    "        while i > 0 and j > 0:\n",
    "            # set trace stop value\n",
    "            if score[i, j] == 0:\n",
    "                break\n",
    "            # match or mismatch\n",
    "            if score[i, j] == score[i-1, j-1] + self.match_score(self.seq1[i-1], self.seq2[j-1]):\n",
    "                align1 += self.seq1[i-1]\n",
    "                align2 += self.seq2[j-1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            # delete\n",
    "            elif score[i, j] == score[i-1, j] + self.gap_penalty:\n",
    "                align1 += self.seq1[i-1]\n",
    "                align2 += '-'\n",
    "                i -= 1\n",
    "            # insert\n",
    "            elif score[i, j] == score[i, j-1] + self.gap_penalty:\n",
    "                align1 += '-'\n",
    "                align2 += self.seq2[j-1]\n",
    "                j -= 1\n",
    "        # Reverse the alignment\n",
    "        align1 = align1[::-1]\n",
    "        align2 = align2[::-1]\n",
    "        return align1, align2\n",
    "\n",
    "    def display_alignment(self):\n",
    "        \"\"\" Display the global alignment between two sequences \"\"\"\n",
    "        # Get the alignment result\n",
    "        align1, align2 = self.traceback()\n",
    "        # Get the map label index\n",
    "        map_label = [\" \"] * len(align1)\n",
    "        for i in range(len(align1)):\n",
    "            if align1[i] == align2[i]:\n",
    "                map_label[i] = \"|\"\n",
    "            elif align1[i] == '-' or align2[i] == '-':\n",
    "                continue\n",
    "            else:\n",
    "                map_label[i] = '.'\n",
    "\n",
    "        map_label = ''.join(map_label)\n",
    "        # Display the alignment\n",
    "        print(align1 + \"\\n\" + map_label + \"\\n\" + align2)\n",
    "\n",
    "    def display_array(self):\n",
    "        \"\"\" Display the score array and trace array \"\"\"\n",
    "        # Get the score array and trace array\n",
    "        score_array, traceback_array = self.score_matrix()\n",
    "        # Display\n",
    "        column_labels = [label for label in \"-\" + \"\".join(self.seq1)]\n",
    "        row_labels = [label for label in \"-\" + \"\".join(self.seq2)]\n",
    "        print('Score Matrix:')\n",
    "        display(self.pretty_table_from_array(score_array, row_labels, column_labels))\n",
    "        print('Traceback Matrix')\n",
    "        display(self.pretty_table_from_array(traceback_array, row_labels, column_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's an example. Let's start with two random unaligned DNA sequences created by `numpy.random`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)\n",
    "x = np.random.choice(['A', 'T', 'G', 'C'], 8)\n",
    "y = np.random.choice(['A', 'T', 'G', 'C'], 8)\n",
    "sw = SmithWaterman(x, y, gap_penalty=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we check out the score matries and traceback matries, firstly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score Matrix:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback Matrix\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>-</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>↑</td>\n",
       "      <td>↑</td>\n",
       "      <td>↑</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↑</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>↑</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>↖</td>\n",
       "      <td>←</td>\n",
       "      <td>↖</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sw.display_array()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we check out the alignment between seq1 and seq2 under different gap penalty:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGCA\n",
      "||||\n",
      "GGCA\n"
     ]
    }
   ],
   "source": [
    "sw.display_alignment()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "281bb314f12105b31675f2dfa5ef5db1db3d4d6a6f8af87c8955db97ce54c5b1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
